File: android/app/src/test/kotlin/io/github/sudomegas/tritium/storage/UnitsTest.kt
package io.github.sudomegas.tritium.storage

import io.github.sudomegas.tritium.storage.Units.ConsumptionUnit
import io.github.sudomegas.tritium.storage.Units.DistanceUnit
import io.github.sudomegas.tritium.storage.Units.VolumeUnit
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class UnitsTest {

    // units.ts's own stated invariant: "Switching to miles and back leaves
    // every file byte-identical, and a test asserts exactly that." Every
    // kilometre reading from 0 to 300,000 — the same range units.ts itself
    // measured the 37.9%-failure-at-zero-decimals claim against — must
    // survive showDistance(MI) then readDistance(MI) unchanged. THE FILE
    // STAYS METRIC is not a comment if nothing checks it.
    @Test
    fun `switching to miles and back leaves the kilometre reading byte-identical`() {
        for (km in 0..300_000) {
            val shown = Units.showDistance(km, DistanceUnit.MI)
            val roundTripped = Units.readDistance(shown, DistanceUnit.MI)
            assertEquals("km=$km round-tripped through miles", km, roundTripped)
        }
    }

    @Test
    fun `kilometres pass through showDistance and readDistance unchanged`() {
        for (km in intArrayOf(0, 1, 19764, 300_000)) {
            assertEquals(km.toLong(), Units.showDistance(km, DistanceUnit.KM))
            assertEquals(km, Units.readDistance(km.toLong(), DistanceUnit.KM))
        }
    }

    @Test
    fun `distance decimals are zero for km and one for mi`() {
        assertEquals(0, Units.DISTANCE_DECIMALS.getValue(DistanceUnit.KM))
        assertEquals(1, Units.DISTANCE_DECIMALS.getValue(DistanceUnit.MI))
    }

    @Test
    fun `19764 km is 12280,8 miles — one decimal, not a whole mile`() {
        // 19764 × 10 / 1.609344 = 122807.9... -> shown as tenths-of-a-mile: 122808
        val shown = Units.showDistance(19764, DistanceUnit.MI)
        assertEquals(122808L, shown)
        assertEquals(19764, Units.readDistance(shown, DistanceUnit.MI))
    }

    @Test
    fun `volume converts to gallons at one extra decimal — exact, like distance`() {
        // 29990 (litres, PUMP_DECIMALS=3, i.e. 29.990 l) -> US gallons at
        // VOLUME_DECIMALS[GAL]=1 extra decimal, so ×10000: 29.990 / 3.785411784
        // = 7.92251... gal -> 79225 at four decimals. The extra digit is
        // exactly DISTANCE_DECIMALS's own move for miles over kilometres,
        // and it buys the same thing: the round trip is exact.
        val gallons = Units.showVolume(29990L, VolumeUnit.GAL)
        assertEquals(79225L, gallons) // round(29990 × 10 / 3.785411784)
        assertEquals(29990L, Units.readVolume(gallons, VolumeUnit.GAL))

        // The same ratio at TANK_DECIMALS (×10, i.e. 54.0 l) — scale-agnostic,
        // per state/units.ts's own tank()/parseTank() reuse of showVolume/readVolume.
        val tankGallons = Units.showVolume(540L, VolumeUnit.GAL)
        assertEquals(1427L, tankGallons) // round(540 × 10 / 3.785411784)
        assertEquals(540L, Units.readVolume(tankGallons, VolumeUnit.GAL))
    }

    @Test
    fun `volume decimals are zero for litres and one extra for gallons`() {
        assertEquals(0, Units.VOLUME_DECIMALS.getValue(VolumeUnit.L))
        assertEquals(1, Units.VOLUME_DECIMALS.getValue(VolumeUnit.GAL))
    }

    @Test
    fun `switching to gallons and back leaves the litre reading byte-identical`() {
        for (litres in 0..100_000 step 7) {
            val shown = Units.showVolume(litres.toLong(), VolumeUnit.GAL)
            val roundTripped = Units.readVolume(shown, VolumeUnit.GAL)
            assertEquals("litres=$litres round-tripped through gallons", litres.toLong(), roundTripped)
        }
    }

    @Test
    fun `litres pass through volume conversion unchanged`() {
        assertEquals(29990L, Units.showVolume(29990L, VolumeUnit.L))
        assertEquals(29990L, Units.readVolume(29990L, VolumeUnit.L))
    }

    @Test
    fun `price per volume multiplies to gallons, divides back to litres`() {
        // A gallon is more litres, so it costs more — units.ts's own warning
        // that this is "the one place it is easy to divide by mistake."
        val perGallon = Units.showPricePerVolume(73380L, VolumeUnit.GAL)
        assertTrue("a price per gallon must be higher than per litre", perGallon > 73380L)
        assertEquals(73380L, Units.readPricePerVolume(perGallon, VolumeUnit.GAL))
    }

    @Test
    fun `showConsumption is identity for l100km, whatever the value`() {
        assertEquals(7000L, Units.showConsumption(7000L, ConsumptionUnit.L100KM))
        assertEquals(0L, Units.showConsumption(0L, ConsumptionUnit.L100KM))
    }

    @Test
    fun `showConsumption returns null at zero for kml and mpg — inverse measures have no image`() {
        assertNull(Units.showConsumption(0L, ConsumptionUnit.KML))
        assertNull(Units.showConsumption(0L, ConsumptionUnit.MPG))
    }

    @Test
    fun `showConsumption converts a real l100km figure to km per litre and mpg`() {
        // 7 l/100km (×1000 = 7000) -> 100/7 = 14.2857 km/l -> ×1000 = 14286
        assertEquals(14286L, Units.showConsumption(7000L, ConsumptionUnit.KML))
        // MPG_CONSTANT (235.214583) / 7 = 33.602... -> ×1000 = 33602
        assertEquals(33602L, Units.showConsumption(7000L, ConsumptionUnit.MPG))
    }

    @Test
    fun `an unrecognised or missing token falls back to the metric default`() {
        assertEquals(DistanceUnit.KM, Units.distanceUnitOf(null))
        assertEquals(DistanceUnit.KM, Units.distanceUnitOf("furlongs"))
        assertEquals(VolumeUnit.L, Units.volumeUnitOf("firkin"))
        assertEquals(ConsumptionUnit.L100KM, Units.consumptionUnitOf("parsecs-per-fortnight"))
    }
}

File: tests/unit/units.test.ts
// The unit boundary (F11), against figures worked by hand.
//
// The claim the whole milestone rests on is that the FILE DOES NOT MOVE: units
// convert on the way to the screen and on the way in from a form, and never in
// between. The round-trip tests are what hold that.

import { describe, expect, it } from 'vitest'
import { draftOf, entryOf } from '../../src/shared/fuel-draft.js'
import type { FuelEntry } from '../../src/shared/records.js'
import { toPump } from '../../src/shared/scaled.js'
import {
  DISTANCE_DECIMALS,
  KM_PER_MILE,
  LITRES_PER_US_GALLON,
  MPG_CONSTANT,
  VOLUME_DECIMALS,
  readDistance,
  readPricePerVolume,
  readVolume,
  showConsumption,
  showCostPerDistance,
  showDistance,
  showPricePerVolume,
  showVolume
} from '../../src/shared/units.js'

describe('the constants', () => {
  it('uses the US gallon and not the imperial one', () => {
    // They differ by twenty per cent, which is the difference between a good
    // tank and a bad one (F11.md decision 3).
    expect(LITRES_PER_US_GALLON).toBe(3.785411784)
    expect(LITRES_PER_US_GALLON).not.toBeCloseTo(4.54609, 5)
  })

  it('uses the exact international mile', () => {
    expect(KM_PER_MILE).toBe(1.609344)
  })

  it('derives the mpg constant from the other two rather than hard-coding it', () => {
    expect(MPG_CONSTANT).toBeCloseTo(235.214583, 5)
  })
})

describe('distance', () => {
  it('shows kilometres unchanged', () => {
    expect(showDistance(19_764, 'km')).toBe(19_764)
    expect(readDistance(19_764, 'km')).toBe(19_764)
  })

  it('shows miles at one decimal, and comes back to the same kilometre', () => {
    // 19.764 km is 12.280,8 mi — held at ×10, which is what the extra decimal
    // buys and why it exists (see DISTANCE_DECIMALS).
    expect(DISTANCE_DECIMALS.mi).toBe(1)
    const miles = showDistance(19_764, 'mi')
    expect(miles).toBe(122_808)
    expect(readDistance(miles, 'mi')).toBe(19_764)
  })

  it('round-trips EVERY kilometre of a realistic odometer range', () => {
    // The property the extra decimal exists for. Whole miles fail this for
    // 37,9% of values, because a mile is coarser than the kilometre under it —
    // a maker in miles would have moved his own odometer by opening a fill-up
    // and saving it untouched.
    for (let km = 0; km <= 300_000; km += 1) {
      expect(readDistance(showDistance(km, 'mi'), 'mi')).toBe(km)
    }
  })

  it('would NOT have round-tripped at whole miles — which is why it does not use them', () => {
    // 3.908 km and 3.907 km both land on 2.428 whole miles, and coming back
    // cannot know which was meant. Pinned so the decimal is never "tidied away".
    const whole = (km: number): number => Math.round(km / KM_PER_MILE)
    expect(whole(3908)).toBe(whole(3907))
    expect(Math.round(whole(3908) * KM_PER_MILE)).not.toBe(3908)
  })
})

describe('volume', () => {
  it('shows litres unchanged', () => {
    expect(showVolume(toPump(54), 'l')).toBe(toPump(54))
  })

  it('converts litres to US gallons, at one extra decimal', () => {
    // 54 l ÷ 3,785411784 = 14,2653 gal — held at PUMP_DECIMALS+1 (×10000),
    // VOLUME_DECIMALS' own extra digit, the same move DISTANCE_DECIMALS
    // makes for miles over kilometres.
    expect(VOLUME_DECIMALS.gal).toBe(1)
    expect(showVolume(toPump(54), 'gal')).toBe(142_653)
  })

  it('round-trips a tank of fuel exactly, the extra decimal is what buys it', () => {
    for (const litres of [1, 10, 29.99, 40, 54, 100]) {
      const scaled = toPump(litres)
      const back = readVolume(showVolume(scaled, 'gal'), 'gal')
      expect(back).toBe(scaled)
    }
  })

  it('round-trips EVERY tenth-of-a-litre of a realistic range', () => {
    for (let scaled = 0; scaled <= 100_000; scaled += 7) {
      expect(readVolume(showVolume(scaled, 'gal'), 'gal')).toBe(scaled)
    }
  })
})

describe('price per volume', () => {
  it('MULTIPLIES going to gallons — a gallon is more litres, so it costs more', () => {
    // The one place it is easy to divide by mistake.
    const perLitre = toPump(73.38)
    const perGallon = showPricePerVolume(perLitre, 'gal')

    expect(perGallon).toBeGreaterThan(perLitre)
    // 73,380 × 3,785411784 = 277,773,5… -> 277.774
    expect(perGallon).toBe(277_774)
  })

  it('round-trips back to the stored per-litre price', () => {
    const perLitre = toPump(73.38)
    const back = readPricePerVolume(showPricePerVolume(perLitre, 'gal'), 'gal')
    expect(Math.abs(back - perLitre)).toBeLessThanOrEqual(1)
  })

  it('leaves a per-litre price alone', () => {
    expect(showPricePerVolume(toPump(73.38), 'l')).toBe(toPump(73.38))
  })
})

describe('consumption', () => {
  const NINE = 9000 // 9,00 l/100km

  it('leaves l/100km alone', () => {
    expect(showConsumption(NINE, 'l100km')).toBe(NINE)
  })

  it('converts to km/l', () => {
    // 100 ÷ 9 = 11,111 km/l
    expect(showConsumption(NINE, 'kml')).toBe(11_111)
  })

  it('converts to mpg', () => {
    // 235,214583 ÷ 9 = 26,135 mpg
    expect(showConsumption(NINE, 'mpg')).toBe(26_135)
  })

  it('has no image for zero in the inverse units', () => {
    // More litres per hundred kilometres is worse; more kilometres per litre is
    // better. Zero has no reciprocal, and infinity is not a figure.
    expect(showConsumption(0, 'kml')).toBe(null)
    expect(showConsumption(0, 'mpg')).toBe(null)
    expect(showConsumption(0, 'l100km')).toBe(0)
  })
})

describe('cost per distance', () => {
  it('MULTIPLIES going to miles — a mile is further, so it costs more', () => {
    expect(showCostPerDistance(1000, 'km')).toBe(1000)
    expect(showCostPerDistance(1000, 'mi')).toBe(1609)
  })
})

describe('the file does not move', () => {
  const ENTRY: FuelEntry = {
    id: 'f-0001',
    date: '2026-08-16',
    odometer_km: 19_764,
    litres: toPump(29.99),
    price_per_litre: toPump(73.38),
    full_tank: true,
    fuel_type: 'Kurşunsuz 95'
  }

  it('shows a fill-up in miles and gallons and stores it in km and litres', () => {
    const imperial = { distance: 'mi' as const, volume: 'gal' as const }
    const draft = draftOf(ENTRY, imperial)

    // What the maker sees: miles to one decimal, gallons to FOUR — the
    // VOLUME_DECIMALS extra digit that makes the round trip exact.
    expect(draft.odometer_km).toBe('12280,8')
    expect(draft.litres).toBe('7,9225')

    // What the file gets back.
    const record = entryOf(draft, imperial)
    expect(record.odometer_km).toBe(19_764)
  })

  it('round-trips a whole record through miles and gallons exactly', () => {
    const imperial = { distance: 'mi' as const, volume: 'gal' as const }
    const back = entryOf(draftOf(ENTRY, imperial), imperial)

    expect(back.odometer_km).toBe(ENTRY.odometer_km)
    expect(back.date).toBe(ENTRY.date)
    expect(back.full_tank).toBe(ENTRY.full_tank)
    // Litres are exact now (VOLUME_DECIMALS's own extra digit). Price per
    // litre carries no extra decimal — see units.ts's own comment on why it
    // does not need one — so it stays within a kuruş the scale can express.
    expect(back.litres).toBe(ENTRY.litres)
    expect(Math.abs(back.price_per_litre - ENTRY.price_per_litre)).toBeLessThanOrEqual(1)
  })

  it('leaves a metric record byte-identical through a metric round trip', () => {
    const metric = { distance: 'km' as const, volume: 'l' as const }
    expect(entryOf(draftOf(ENTRY, metric), metric)).toEqual({
      date: ENTRY.date,
      odometer_km: ENTRY.odometer_km,
      litres: ENTRY.litres,
      price_per_litre: ENTRY.price_per_litre,
      full_tank: ENTRY.full_tank,
      fuel_type: ENTRY.fuel_type
    })
  })

  it('keeps the total the same in either unit, because the conversions cancel', () => {
    // Gallons × price-per-gallon is the same money as litres × price-per-litre,
    // which is why the field pair must always move together (F11.md decision 5).
    const litres = showVolume(toPump(29.99), 'l')
    const perLitre = showPricePerVolume(toPump(73.38), 'l')
    const gallons = showVolume(toPump(29.99), 'gal')
    const perGallon = showPricePerVolume(toPump(73.38), 'gal')

    const metricTotal = (litres * perLitre) / 1_000_000
    // gallons carries VOLUME_DECIMALS.gal's own extra decimal (×10000, not
    // ×1000), so the product needs one more zero to land back on money.
    const imperialTotal = (gallons * perGallon) / 10_000_000

    expect(Math.abs(metricTotal - imperialTotal)).toBeLessThan(0.5)
  })
})
